python/ray/tune/automl/genetic_searcher.py

```diff
@@ -0,0 +1,206 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# author: yiguang.wyg (yiguang.wyg@antfin.com)
+# date: 2018/06/27
+
+import logging
+import numpy as np
+
+from .search_policy import AutoMLSearcher
+
+
+class GeneticSearch(AutoMLSearcher):
+    """Implement the genetic search.
+
+    Keep a collection of top-K parameter permutations as base genes,
+    then apply selection, crossover, and mutation to them to generate
+    new genes (a.k.a new generation). Hopefully, the performance of
+    the top population would increase generation by generation.
+    """
+
+    def __init__(self, search_space, reward_attr,
+                 max_generation=2,
+                 population_size=10,
+                 keep_top_ratio=0.2,
+                 selection_bound=0.4,
+                 crossover_bound=0.6):
+        """
+        Initialize GeneticSearcher.
+
+        Args:
+            search_space (SearchSpace): The space to search.
+            reward_attr: The attribute name of the reward in the result.
+            max_generation: Max iteration number of genetic search.
+            population_size: Number of trials of the initial generation.
+            keep_top_ratio: Ratio of the top performance population.
+            selection_bound: Threshold for performing selection.
+            crossover_bound: Threshold for performing crossover.
+        """
+        super(GeneticSearch, self).__init__(search_space, reward_attr)
+
+        self._cur_generation = 1
+        self._max_generation = max_generation
+        self._population_size = population_size
+        self._keep_top_ratio = keep_top_ratio
+        self._selection_bound = selection_bound
+        self._crossover_bound = crossover_bound
+
+        self._cur_config_list = []
+        self._cur_encoding_list = []
+        for _ in range(population_size):
+            one_hot = self.search_space.generate_random_one_hot_encoding()
+            self._cur_encoding_list.append(one_hot)
+            self._cur_config_list.append(
+                self.search_space.apply_one_hot_encoding(one_hot))
+
+    def _select(self):
+        population_size = len(self._cur_config_list)
+        logging.info("The %sth generation, population=%s"
```

Comment:
In every module (Python fie). At the top of every module you want to log something, create `logger = logging.getLogger(__name__) ` and then use `logger` instance.

Reply:
Also (while I disagree), @robertnishihara pointed out that he would prefer all interpolations to currently use `{}` format interpolation, and not old-school `%` interpolation.

BTW, as a side note, if you do use `% ` interpolation, which is native to logging, then you should not do interpolation here yourself, but pass it to logging to interpolate if needed (it might be that the message is discarded and thus you do not pay the cost of interpolation).

So:

```python
logger.info(The %sth generation, population=%s", self._cur_generation, population_size)
```

Personally, I prefer a bit more verbose style:

```python
logger.info(The %(current_generation)sth generation, population=%(population_size)s", {'current_generation': self._cur_generation, 'population_size': population_size})
```